minimint-core/src/epoch.rs

```diff
@@ -0,0 +1,227 @@
+use minimint_derive::UnzipConsensus;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+pub type ConsensusOutcome = Batch<Vec<ConsensusItem>, PeerId>;
+
+#[derive(
+    Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize, UnzipConsensus, Encodable, Decodable,
+)]
+pub enum ConsensusItem {
+    EpochInfo(EpochSignatureShare),
+    Transaction(Transaction),
+    Mint(<minimint_mint::Mint as FederationModule>::ConsensusItem),
+    Wallet(<minimint_wallet::Wallet as FederationModule>::ConsensusItem),
+    LN(<minimint_ln::LightningModule as FederationModule>::ConsensusItem),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct EpochSignatureShare(pub SignatureShare);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub struct EpochSignature(pub Signature);
+
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable)]
+pub struct EpochHistory {
+    pub outcome: MerkleOutcome,
+    pub hash: Sha256,
+    pub signature: Option<EpochSignature>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable)]
+pub struct MerkleOutcome {
+    pub epoch: u64,
+    pub last_hash: Option<Sha256>,
+    pub items: Vec<(PeerId, Vec<ConsensusItem>)>,
+}
+
+impl MerkleOutcome {
+    pub fn hash(&self) -> Sha256 {
+        let mut bytes: Vec<u8> = vec![];
+        self.consensus_encode(&mut bytes).unwrap();
+        Sha256::hash(&bytes)
+    }
+
+    pub fn to_hbbft(self) -> ConsensusOutcome {
+        Batch {
+            epoch: self.epoch,
+            contributions: BTreeMap::from_iter(self.items.into_iter()),
+        }
+    }
+}
+
+impl EpochHistory {
+    pub fn new(
+        epoch: u64,
+        contributions: BTreeMap<PeerId, Vec<ConsensusItem>>,
+        prev_epoch: &Option<EpochHistory>,
+    ) -> Self {
+        let items = contributions
+            .into_iter()
+            .sorted_by_key(|(peer, _)| *peer)
+            .collect();
+        let outcome = MerkleOutcome {
+            last_hash: prev_epoch.clone().map(|epoch| epoch.hash),
+            items,
+            epoch,
+        };
+
+        EpochHistory {
+            hash: outcome.hash(),
+            outcome,
+            signature: None,
+        }
+    }
+
+    /// verifies the sig, the hash, and the merkle tree
+    /// callers must check whether the sig exists
+    pub fn verify(&self, pks: &PublicKey, prev_epoch: Option<&EpochHistory>) -> bool {
+        if let Some(sig) = &self.signature {
+            if !pks.verify(&sig.0, self.hash) {
+                return false;
+            }
+        }
+
+        if self.outcome.epoch > 0 {
+            match prev_epoch {
+                None => return false,
+                Some(epoch) if Some(epoch.outcome.hash()) != self.outcome.last_hash => {
+                    return false
+                }
+                _ => {}
+            }
+        }
+
+        self.hash == self.outcome.hash()
+    }
```

Comment:
nit: Some more logging for the failure cases would be nice. We might also want to put some asserts in place for things that should never happen (like 0th epoch having a prev_epoch supplied).